keps/0017-pipe-tasks.md

```diff
@@ -0,0 +1,122 @@
+---
+kep-number: 17
+title: Pipe Tasks
+authors:
+  - "@zen-dog"
+owners:
+  - "@zen-dog"
+editor: TBD
+creation-date: 2019-09-12
+last-updated: 2019-09-12
+status: provisional
+---
+
+# Pipe Tasks
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Limitations](#limitations)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints-optional)
+* [Alternatives](#alternatives)
+
+## Summary
+
+Developing complicated operators often require generating files in one task and reusing them in a later one. A common example is generating custom certificates/dynamic configuration files in the bootstrap task and using them in the later deployment step of the service. This KEP describes how KUDO could help operator developers by making this task seamless and automated.
+
+## Motivation
+
+We want to ease the development of operators with complicated life-cycles by enabling to pipe files from one task to another. While this can be achieved in a manual way, involving custom scripts, this is exactly the kind of "magic glue" that KUDO could provide to the operator developers to ease the development process.
+
+### Goals
+
+The goal of this KEP is to describe application examples, limitations and ways to implement passing files between tasks. We aim to help operator developers pipe small files (<1Mb) like certificates and dynamically generated configuration which can be referenced later as ConfigMaps and Secrets.
+
+### Non-Goals
+
+Allowing to pipe all kind of files (>1Mb) between tasks requires a general-purpose storage solution (e.g. S3) available in the cluster. While certainly possible, this is not the goal of this KEP.
+
+## Proposal
+
+This section describes how pipe tasks and files they produce are configured in the operator. Here is a task definition that produces a file that will be stored as a Secret, referenced by `{{.Pipes.Certificate}}` key:
+```yaml
+steps:
+    - name: bootstrap
+    tasks:
+        - gencert
+            - pipe:
+                - file: /usr/share/MyKey.key
+                  kind: Secret # ConfigMap
+                  key: {{.Pipes.Certificate}}
+``` 
+The key `{{.Pipes.Certificate}}` value will be generated by KUDO to avoid collisions and to make sure that it is not used before the file is generated. In the above example we would generate a secret name like `deploy.bootstrap.gencert.pipes.certificate-#hash` to capture which plan/phase/step/task it belongs too as well as the hash of its content. KUDO would use this hash to avoid recreating secrets for unchanged files on a plan rerun. We would also use labels ensuring that the secret is cleaned up when the corresponding Instance is removed.
```

Comment:
Do we need instance name in that or is it ok for multiple instances sharing the same secret, that was maybe generated by another instance 🤔 I guess the secret will be namespaced so if both instances are running in the same namespace, it's ok to share? 🤔 dunno, just thinking out loud at this point :)

Reply:
Good point. I would not share secrets between Instances since this requires coordinating secret life-cycles (e.g. Instance is deleted but Secret should stay because it is used by another Instance). I'll add `Instance.Name` into the name.